File: api/analyze.js
const express = require('express');
const AWS = require('aws-sdk');
const { exec } = require('child_process');
const fs = require('fs');
const path = require('path');

const router = express.Router();

AWS.config.update({
  accessKeyId: process.env.AWS_ACCESS_KEY_ID,
  secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
  region: process.env.AWS_REGION
});

const s3 = new AWS.S3();

router.post('/analyze', async (req, res) => {
  try {
    const { video_url } = req.body;
    if (!video_url) {
      return res.status(400).json({ error: 'URL do vídeo é obrigatória' });
    }

    const filePath = '/tmp/video.mp4';
    const command = `yt-dlp -o "${filePath}" "${video_url}"`;

    exec(command, async (error, stdout, stderr) => {
      if (error) {
        console.error('Erro ao baixar vídeo:', error);
        return res.status(500).json({ error: 'Falha ao baixar o vídeo' });
      }

      const fileContent = fs.readFileSync(filePath);

      const params = {
        Bucket: process.env.AWS_S3_BUCKET_NAME,
        Key: 'videos/video.mp4',
        Body: fileContent
      };

      try {
        await s3.upload(params).promise();
        const url = s3.getSignedUrl('getObject', {
          Bucket: process.env.AWS_S3_BUCKET_NAME,
          Key: 'videos/video.mp4',
          Expires: 3600
        });

        return res.status(200).json({ url });
      } catch (uploadError) {
        console.error('Erro no upload:', uploadError);
        return res.status(500).json({ error: 'Falha ao enviar para o S3' });
      }
    });
  } catch (err) {
    console.error('Erro geral:', err);
    return res.status(500).json({ error: 'Erro interno do servidor' });
  }
});

module.exports = router;

File: api/analyze.py
from fastapi import FastAPI
import yt_dlp, boto3, os

app = FastAPI()

# Cliente S3 usando variáveis de ambiente
s3 = boto3.client(
    "s3",
    aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID"),
    aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY"),
)
BUCKET = os.getenv("S3_BUCKET", "gptvideosbaile")

# Health-check em /
@app.get("/")
def health():
    return {"status":"alive"}

# Seu endpoint principal
@app.post("/analyze")
def analyze(payload: dict):
    video_url = payload.get("video_url")
    if not video_url:
        return {"error":"payload precisa de video_url"}

    # 1) Baixar vídeo para /tmp/video.mp4
    ydl_opts = {"outtmpl":"/tmp/video.mp4"}
    with yt_dlp.YoutubeDL(ydl_opts) as ydl:
        ydl.download([video_url])

    # 2) Enviar ao S3
    key = f"videos/{os.path.basename('/tmp/video.mp4')}"
    s3.upload_file("/tmp/video.mp4", BUCKET, key)

    # 3) Gerar link pré-assinado
    url = s3.generate_presigned_url(
        "get_object",
        Params={"Bucket":BUCKET,"Key":key},
        ExpiresIn=3600
    )
    return {"url":url}